src/VoipUserMapper.ts

```diff
@@ -17,63 +17,96 @@ limitations under the License.
+import RoomListStore from './stores/room-list/RoomListStore';
+import { Room } from 'matrix-js-sdk/src/models/room';
 
-// Functions for mapping users & rooms for the voip_mxid_translate_pattern
-// config option
+// Functions for mapping virtual users & rooms. Currently the only lookup
+// is sip virtual: there could be others in the future.
 
-export function voipUserMapperEnabled(): boolean {
-    return SdkConfig.get()['voip_mxid_translate_pattern'] !== undefined;
-}
-
-// only exported for tests
-export function userToVirtualUser(userId: string, templateString?: string): string {
-    if (templateString === undefined) templateString = SdkConfig.get()['voip_mxid_translate_pattern'];
-    if (!templateString) return null;
-    return templateString.replace('${mxid}', encodeURIComponent(userId).replace(/%/g, '=').toLowerCase());
-}
-
-// only exported for tests
-export function virtualUserToUser(userId: string, templateString?: string): string {
-    if (templateString === undefined) templateString = SdkConfig.get()['voip_mxid_translate_pattern'];
-    if (!templateString) return null;
+export default class VoipUserMapper {
+    private virtualRoomIdCache = new Set<string>();
 
-    const regexString = templateString.replace('${mxid}', '(.+)');
-
-    const match = userId.match('^' + regexString + '$');
-    if (!match) return null;
+    public static sharedInstance(): VoipUserMapper {
+        if (window.mxVoipUserMapper === undefined) window.mxVoipUserMapper = new VoipUserMapper();
+        return window.mxVoipUserMapper;
+    }
 
-    return decodeURIComponent(match[1].replace(/=/g, '%'));
-}
+    private async userToVirtualUser(userId: string): Promise<string> {
+        const results = await CallHandler.sharedInstance().sipVirtualLookup(userId);
+        if (results.length === 0) return null;
+        return results[0].userid;
+    }
 
-async function getOrCreateVirtualRoomForUser(userId: string):Promise<string> {
-    const virtualUser = userToVirtualUser(userId);
-    if (!virtualUser) return null;
+    public async getOrCreateVirtualRoomForRoom(roomId: string):Promise<string> {
+        const userId = DMRoomMap.shared().getUserIdForRoomId(roomId);
+        if (!userId) return null;
+
+        const virtualUser = await this.userToVirtualUser(userId);
+        if (!virtualUser) return null;
+
+        // There's quite a bit of acromatics here to prevent the virtual room being shown
+        // while it's being created: forstly, we have to stop the RoomListStore from showing
+        // new rooms for a bit, because we can't set the room account data to say it's a virtual
+        // room until we have the room ID. Secondly, once we have the new room ID, we have to
+        // temporarily cache the fact it's a virtual room because there's no local echo on
+        // room account data so it won't show up in the room model until it comes down the
+        // sync stream again. Ick.
```

Comment:
Hmm. I agree this is unfortunate... Instead of holding _all_ new rooms because of just one, perhaps an API where you hold just room X by ID? Even if it's out of scope here, what changes would avoid the need for such hackery?

Reply:
Unfortunately that's the problem: we don't know what the room ID is before the API call has returned, so if we wait for that the room could already have come down the sync :(

Reply:
Hmm, I'm really not that fond of holding _all_ rooms like this, since it could go wrong in unexpected ways:

1. Perhaps it's not possible to have two calls at once, but if it were, you could starting holding for room A, then start and stop holding for room B, but it's global, so that means you stop holding for all rooms
2. If `ensureDMExists` ever wedges such that it never finishes, then suddenly we're stuck in a surprising mode where we never show new rooms

Let's try to think of a different approach... So you know the virtual user up front, but not the virtual room ID? Could the room list store check each new room with some module that knows who the virtual users are...? 

There must be _some_ approach that doesn't require a locking out unrelated rooms as collateral damage.

Reply:
Yeah, we could guess the virtual-ness of rooms by looking for rooms with us and exactly one other member that's virtual. I've tried this, but it's not working as it seems like the other member's join event isn't coming down until another sync happens so the room still starts off as visible still (also explains why whenever you create a room, it's briefly called 'Empty Room'). Once the room is visible, it never hides until you refresh because the room list assumes room visibility doesn't change (it just suppresses updates for invisible rooms).

This might still be a viable option but we'll have to change the room list store so room visibility can change and/or fix synapse so all the initial room events come down in one sync. Does this seem plausible before I start doing one or both of those?

Reply:
Right, have tried signalling the virtual-ness via room creation content which is feeling a lot nicer.

Reply:
Thanks, the creation content approach is much nicer! 😄